crowbar_framework/lib/crowbar/upgrade_status.rb

```diff
@@ -92,17 +92,22 @@ def start_step(step_name, out_of_order = false)
       end
     end
 
-    def end_step(success = true, errors = {})
+    def end_step(step_name = nil, out_of_order = false, success = true, errors = {})
       ::Crowbar::Lock::LocalBlocking.with_lock(shared: false, logger: @logger, path: lock_path) do
-        unless running?
+        step_name ||= current_step
+        unless running? step_name
           @logger.warn("The step is not running, could not be finished")
           return false
         end
-        progress[:steps][current_step] = {
+        if step_name != current_step && !out_of_order
```

Comment:
This one is tricky. Running steps "out of order" should be called rather something like running steps in "detached" mode. This is meant for steps like prechecks or repochecks which should be callable any time even when some other step is running.
With this implementation, this would cause the status to show those "detached" steps as "running" for a short while which might cause some unwanted side effects.